File: desktop/bin/fronte.swift
// myynd-fronte: chi c'è davanti sul Mac, una riga JSON a ogni cambio.
//
// Lo lancia il guscio (`desktop/fronte.ts`) e lo tiene in vita col suo stdin:
// quando stdin si chiude — Myynd è uscita, anche male — il programma esce.
// Non chiede mai permessi. Con `--titoli` legge il titolo della finestra
// davanti solo se il Mac ha già dato a Myynd il permesso di Accessibilità;
// senza, il titolo è sempre null.
//
//   myynd-fronte --versione   → {"versione":1} ed esce (l'unico modo di provarlo)
//   myynd-fronte [--titoli]   → {"bundle":"com.apple.Safari","app":"Safari","pid":123,"titolo":"…","t":1727160000000}
//
// Su stdin, una riga «ora» fa scrivere subito chi c'è davanti, anche se non è
// cambiato: il guscio la manda dopo un blocco dello schermo o un sonno.
//
// Lo costruisce `build/fronte.cjs` (universale, arm64 e x86_64).

import AppKit
import ApplicationServices

setvbuf(stdout, nil, _IOLBF, 0)

let argomenti = CommandLine.arguments
if argomenti.contains("--versione") {
  print("{\"versione\":1}")
  exit(0)
}
let conTitoli = argomenti.contains("--titoli")

let TITOLO_MASSIMO = 300
let RINVIO = 0.3
let OGNI = 5.0

/** Il permesso di Accessibilità, senza mai chiederlo. */
func fidato() -> Bool {
  let opzioni = [kAXTrustedCheckOptionPrompt.takeUnretainedValue() as String: false] as CFDictionary
  return AXIsProcessTrustedWithOptions(opzioni)
}

func scrivi(_ oggetto: [String: Any]) {
  guard let dati = try? JSONSerialization.data(withJSONObject: oggetto, options: [.withoutEscapingSlashes]) else { return }
  FileHandle.standardOutput.write(dati)
  FileHandle.standardOutput.write("\n".data(using: .utf8)!)
}

final class Osservatore {
  var pid: pid_t = 0
  var elementoApp: AXUIElement?
  var finestra: AXUIElement?
  var osservatore: AXObserver?
  var ultima = ""
  var rinvio: DispatchWorkItem?

  init() {
    let centro = NSWorkspace.shared.notificationCenter
    centro.addObserver(forName: NSWorkspace.didActivateApplicationNotification, object: nil, queue: .main) { [weak self] nota in
      let app = nota.userInfo?[NSWorkspace.applicationUserInfoKey] as? NSRunningApplication
      self?.attiva(app)
    }
    Timer.scheduledTimer(withTimeInterval: OGNI, repeats: true) { [weak self] _ in
      self?.emetti(sempre: false)
    }
    attiva(NSWorkspace.shared.frontmostApplication)
    emetti(sempre: true)
  }

  /** Un'app nuova davanti: ci si iscrive ai suoi cambi di finestra e di titolo. */
  func attiva(_ app: NSRunningApplication?) {
    togliOsservatore()
    pid = app?.processIdentifier ?? 0
    if app != nil, conTitoli, fidato() {
      let el = AXUIElementCreateApplication(pid)
      AXUIElementSetMessagingTimeout(el, 0.5)
      elementoApp = el
      var obs: AXObserver?
      let richiamo: AXObserverCallback = { _, _, _, refcon in
        guard let refcon = refcon else { return }
        let me = Unmanaged<Osservatore>.fromOpaque(refcon).takeUnretainedValue()
        me.cambiato()
      }
      if AXObserverCreate(pid, richiamo, &obs) == .success, let obs = obs {
        osservatore = obs
        let io = Unmanaged.passUnretained(self).toOpaque()
        AXObserverAddNotification(obs, el, kAXFocusedWindowChangedNotification as CFString, io)
        AXObserverAddNotification(obs, el, kAXTitleChangedNotification as CFString, io)
        CFRunLoopAddSource(CFRunLoopGetMain(), AXObserverGetRunLoopSource(obs), .defaultMode)
        seguiFinestra()
      }
    }
    pianifica()
  }

  func togliOsservatore() {
    if let obs = osservatore {
      if let el = elementoApp {
        AXObserverRemoveNotification(obs, el, kAXFocusedWindowChangedNotification as CFString)
        AXObserverRemoveNotification(obs, el, kAXTitleChangedNotification as CFString)
      }
      if let w = finestra { AXObserverRemoveNotification(obs, w, kAXTitleChangedNotification as CFString) }
      CFRunLoopRemoveSource(CFRunLoopGetMain(), AXObserverGetRunLoopSource(obs), .defaultMode)
    }
    osservatore = nil
    elementoApp = nil
    finestra = nil
  }

  /** Il titolo cambia sulla finestra, non sull'app: ci si iscrive a quella davanti. */
  func seguiFinestra() {
    guard let obs = osservatore else { return }
    if let w = finestra { AXObserverRemoveNotification(obs, w, kAXTitleChangedNotification as CFString) }
    finestra = finestraDavanti()
    if let w = finestra {
      AXObserverAddNotification(obs, w, kAXTitleChangedNotification as CFString, Unmanaged.passUnretained(self).toOpaque())
    }
  }

  func cambiato() {
    seguiFinestra()
    pianifica()
  }

  /** I cambi arrivano a raffiche: si scrive trecento millesimi dopo l'ultimo. */
  func pianifica() {
    rinvio?.cancel()
    let lavoro = DispatchWorkItem { [weak self] in self?.emetti(sempre: false) }
    rinvio = lavoro
    DispatchQueue.main.asyncAfter(deadline: .now() + RINVIO, execute: lavoro)
  }

  func finestraDavanti() -> AXUIElement? {
    guard let el = elementoApp else { return nil }
    var valore: CFTypeRef?
    guard AXUIElementCopyAttributeValue(el, kAXFocusedWindowAttribute as CFString, &valore) == .success,
          let v = valore, CFGetTypeID(v) == AXUIElementGetTypeID() else { return nil }
    return (v as! AXUIElement)
  }

  func titolo() -> String? {
    guard conTitoli, fidato(), let w = finestraDavanti() else { return nil }
    var valore: CFTypeRef?
    guard AXUIElementCopyAttributeValue(w, kAXTitleAttribute as CFString, &valore) == .success,
          let s = valore as? String, !s.isEmpty else { return nil }
    return String(s.prefix(TITOLO_MASSIMO))
  }

  /** Chi c'è davanti adesso; si scrive solo se è cambiato, o se lo si chiede. */
  func emetti(sempre: Bool) {
    guard let app = NSWorkspace.shared.frontmostApplication else { return }
    if app.processIdentifier != pid && elementoApp != nil { attiva(app); return }
    let bundle = app.bundleIdentifier ?? ""
    let nome = app.localizedName ?? bundle
    let t = titolo()
    let chiave = "\(bundle)\u{0}\(app.processIdentifier)\u{0}\(t ?? "\u{1}")"
    if !sempre && chiave == ultima { return }
    ultima = chiave
    scrivi([
      "bundle": bundle, "app": nome, "pid": Int(app.processIdentifier),
      "titolo": t.map { $0 as Any } ?? NSNull(),
      "t": Int64((Date().timeIntervalSince1970 * 1000).rounded())
    ])
  }
}

let io = Osservatore()

// stdin chiuso: il guscio non c'è più, e nemmeno noi
Thread.detachNewThread {
  while let riga = readLine(strippingNewline: true) {
    if riga == "ora" { DispatchQueue.main.async { io.emetti(sempre: true) } }
  }
  exit(0)
}

RunLoop.main.run()

File: desktop/fronte.ts
// Chi c'è davanti: le due fonti dell'osservatore.
//
// La prima è un piccolo programma nostro, `bin/myynd-fronte` (il sorgente è
// `bin/fronte.swift`, lo costruisce `build/fronte.cjs`): ascolta il Mac
// invece di chiedergli ogni tanto, e scrive una riga JSON a ogni cambio — app
// davanti e, se la persona ha dato il permesso di Accessibilità, titolo della
// finestra. Muore da solo quando il guscio chiude il suo stdin, cioè quando
// Myynd esce, anche se esce male.
//
// La seconda è il ripiego, quando il programma non c'è (in sviluppo, prima di
// `npm run fronte`) o continua a cadere: `lsappinfo`, che il Mac ha di serie,
// chiesto ogni cinque secondi. Solo l'app, mai il titolo.
//
// Le due letture delle righe sono pure e provate su esempi veri. Le fonti vere
// non girano mai in una prova: guarderebbero la persona che sta lavorando.

import { spawn, execFile } from 'node:child_process'
import { accessSync, constants } from 'node:fs'
import { createInterface } from 'node:readline'
import { fileURLToPath } from 'node:url'
import type { EventoFronte } from './sessioni.ts'

export type FonteFronte = {
  /** Da dove arrivano gli eventi: il programma nostro, o `lsappinfo`. */
  tipo: 'aiutante' | 'ripiego'
  /** Chi c'è davanti adesso, anche se non è cambiato (dopo un blocco, un sonno). */
  chiedi(): void
  ferma(): void
}
export type OpzioniFonte = {
  titoli: boolean
  suEvento: (e: EventoFronte) => void
  /** La fonte si è fermata da sola (il programma è uscito, non è partito). */
  suFine: (motivo: string) => void
  /** Solo il ripiego, anche se il programma c'è: dopo tre cadute. */
  ripiego?: boolean
}
export type Crea = (o: OpzioniFonte) => FonteFronte

/** Il segno di direzione che macOS mette davanti al nome di certe app, e i suoi fratelli. */
const DIREZIONE = /[‎‏‪-‮]/g

/* ------------------------------------------------------------------ letture */

/** Una riga del programma: quella forma e nient'altro. */
export function leggiRiga(riga: string): EventoFronte | null {
  let v: unknown
  try { v = JSON.parse(riga) } catch { return null }
  if (!v || typeof v !== 'object' || Array.isArray(v)) return null
  const { bundle, app, pid, titolo, t } = v as Record<string, unknown>
  if (typeof bundle !== 'string' || typeof app !== 'string') return null
  if (typeof pid !== 'number' || !Number.isInteger(pid)) return null
  if (titolo !== null && typeof titolo !== 'string') return null
  if (typeof t !== 'number' || !Number.isFinite(t)) return null
  return { bundle, app, pid, titolo, t }
}

/** L'ASN che `lsappinfo front` scrive: `ASN:0x0-0x1d01d:`. */
export function leggiAsn(testo: string): string | null {
  const m = /ASN:0x[0-9a-f]+-0x[0-9a-f]+:?/i.exec(testo)
  return m ? m[0] : null
}

/**
 * Quello che `lsappinfo info -only name -only bundleid -only pid <ASN>` scrive:
 *
 *   "LSDisplayName"="‎WhatsApp"
 *   "CFBundleIdentifier"="net.whatsapp.WhatsApp"
 *   "pid"=75956
 *
 * Senza bundle non si sa chi è: `null`.
 */
export function leggiLsappinfo(testo: string, t: number): EventoFronte | null {
  const bundle = /"CFBundleIdentifier"="([^"]*)"/.exec(testo)?.[1]?.replace(DIREZIONE, '').trim()
  if (!bundle) return null
  const nome = /"LSDisplayName"="([^"]*)"/.exec(testo)?.[1]?.replace(DIREZIONE, '').trim()
  const pid = Number(/"pid"\s*=\s*(\d+)/.exec(testo)?.[1] ?? 0)
  return { bundle, app: nome || bundle, pid: Number.isInteger(pid) ? pid : 0, titolo: null, t }
}

/* ---------------------------------------------------------------- percorso */

/** Dentro l'asar un eseguibile non parte: quello vero sta in `app.asar.unpacked`. */
export function spacchettato(percorso: string): string {
  return percorso.replace(/([\\/])app\.asar([\\/])/, '$1app.asar.unpacked$2')
}

function eseguibile(p: string): boolean {
  try { accessSync(p, constants.X_OK); return true } catch { return false }
}

/** Dov'è il programma, o `null` se non c'è o non si può eseguire. */
export function percorsoHelper(o: { url?: URL; esiste?: (p: string) => boolean } = {}): string | null {
  const url = o.url ?? new URL('./bin/myynd-fronte', import.meta.url)
  const p = spacchettato(fileURLToPath(url))
  return (o.esiste ?? eseguibile)(p) ? p : null
}

/* -------------------------------------------------------------------- fonti */

/** Il programma nostro: un evento per riga, finché il suo stdin resta aperto. */
function creaAiutante(percorso: string, o: OpzioniFonte): FonteFronte {
  let finita = false
  const figlio = spawn(percorso, o.titoli ? ['--titoli'] : [], { stdio: ['pipe', 'pipe', 'ignore'] })
  const fine = (motivo: string) => {
    if (finita) return
    finita = true
    o.suFine(motivo)
  }
  figlio.on('error', () => fine('errore'))
  figlio.on('exit', () => fine('uscito'))
  // uno stdin rotto (il programma è già uscito) non deve far cadere il guscio
  figlio.stdin?.on('error', () => {})
  if (figlio.stdout) {
    const righe = createInterface({ input: figlio.stdout })
    righe.on('line', riga => {
      if (finita) return
      const e = leggiRiga(riga)
      if (e) o.suEvento(e)
    })
  }
  return {
    tipo: 'aiutante',
    chiedi() {
      if (finita) return
      try { figlio.stdin?.write('ora\n') } catch { /* è già uscito */ }
    },
    ferma() {
      finita = true
      try { figlio.stdin?.end() } catch { /* già chiuso */ }
      try { figlio.kill('SIGTERM') } catch { /* già andato */ }
    }
  }
}

const OGNI_RIPIEGO = 5_000

function eseguiVero(cmd: string, argomenti: string[]): Promise<string> {
  return new Promise(risolvi => {
    execFile(cmd, argomenti, { timeout: 3_000, maxBuffer: 64 * 1024 }, (errore, stdout) => {
      risolvi(errore ? '' : String(stdout ?? ''))
    })
  })
}

/**
 * `lsappinfo` ogni cinque secondi: solo l'app, e solo quando cambia.
 * `esegui` si cambia solo nelle prove.
 */
export function creaRipiego(o: OpzioniFonte, esegui: (cmd: string, argomenti: string[]) => Promise<string> = eseguiVero): FonteFronte {
  let fermata = false
  let ultimo = ''
  let inCorso = false
  // un chiedi() arrivato mentre un giro è a metà: si rifà appena finisce
  let richiesto = false
  const giro = async (sempre: boolean) => {
    if (fermata) return
    if (inCorso) { if (sempre) richiesto = true; return }
    inCorso = true
    try {
      const asn = leggiAsn(await esegui('lsappinfo', ['front']))
      if (!asn || fermata) return
      const e = leggiLsappinfo(await esegui('lsappinfo', ['info', '-only', 'name', '-only', 'bundleid', '-only', 'pid', asn]), Date.now())
      if (!e || fermata) return
      const chiave = `${e.bundle}\u0000${e.pid}`
      if (!sempre && chiave === ultimo) return
      ultimo = chiave
      o.suEvento(e)
    } finally {
      inCorso = false
      if (richiesto && !fermata) { richiesto = false; void giro(true) }
    }
  }
  const orologio = setInterval(() => { void giro(false) }, OGNI_RIPIEGO)
  orologio.unref?.()
  void giro(true)
  return {
    tipo: 'ripiego',
    chiedi() { void giro(true) },
    ferma() { fermata = true; clearInterval(orologio) }
  }
}

/** Il programma se c'è e parte; altrimenti, o se lo si chiede, il ripiego. */
export const creaFonte: Crea = o => {
  const p = o.ripiego ? null : percorsoHelper()
  if (p) {
    try { return creaAiutante(p, o) } catch { /* non è partito: ripiego */ }
  }
  return creaRipiego(o)
}
